binaries/cuprated/src/tor.rs

```diff
@@ -0,0 +1,190 @@
+}
+
+/// Initialize the Tor network if enabled in configuration
+///
+/// This function will bootstrap Arti if needed by Tor network zone or
+/// clearnet as a proxy.
+pub async fn initialize_tor_if_enabled(config: &Config) -> TorContext {
+    let mode = config.tor.mode;
+    let anonymize_clearnet = config.p2p.clear_net.proxy.to_lowercase() == "Tor";
+
+    // Start Arti client
+    let (bootstrapped_client, arti_client_config) =
+        if mode == TorMode::Arti && ( config.p2p.tor_net.enabled || anonymize_clearnet ) {
+            Some(initialize_arti_client(config).await)
+        } else {
+            None
+        }
+        .unzip();
+
+    // Start Arti onion service
+    let arti_onion_service = arti_client_config
+        .as_ref()
+        .map(|client_config| initialize_arti_onion_service(client_config, config));
+
+    TorContext {
+        mode,
+        bootstrapped_client,
+        arti_client_config,
+        arti_onion_service,
+    }
+}
+
+/// Initialize Arti Tor client.
+async fn initialize_arti_client(config: &Config) -> (TorClient<PreferredRuntime>, TorClientConfig) {
+    // Configuration
+    let mut tor_config = TorClientConfig::builder();
+
+    // Storage
+    tor_config
+        .storage()
+        .state_dir(CfgPath::new_literal(config.tor.arti.directory_path.clone()));
+
+    let tor_config = tor_config
+        .build()
+        .expect("Failed to build Tor client configuration.");
+
+    // Bootstrapping
+    info!("Bootstrapping Arti's TorClient...");
+    let mut tor_client = TorClient::builder()
+        .config(tor_config.clone())
+        .create_bootstrapped()
+        .await
+        .inspect_err(|err| tracing::error!("Unable to bootstrap arti: {err}"))
+        .unwrap();
+
+    // Isolation
+    if config.tor.arti.isolated_circuit {
+        let mut stream_prefs = StreamPrefs::new();
+        stream_prefs.isolate_every_stream();
+        tor_client.set_stream_prefs(stream_prefs);
+    }
+
+    (tor_client, tor_config)
+}
+
+fn initialize_arti_onion_service(client_config: &TorClientConfig, config: &Config) -> OnionService {
+    let onion_svc_config = OnionServiceConfigBuilder::default()
+        .enable_pow(config.tor.arti.onion_service_pow)
+        .nickname(HsNickname::new("cuprate".into()).unwrap())
+        .build()
+        .unwrap();
+
+    TorClient::<PreferredRuntime>::create_onion_service(client_config, onion_svc_config)
+        .expect("Unable to start Arti onion service.")
+}
+
+//---------------------------------------------------------------------------------------------------- Transport configuration
+
+pub fn transport_arti_config(config: &Config, ctx: TorContext) -> TransportConfig<Tor, Arti> {
+    // Extracting
+    let Some(bootstrapped_client) = ctx.bootstrapped_client else {
+        panic!("Arti client should be initialized");
+    };
+    let Some(client_config) = ctx.arti_client_config else {
+        panic!("Arti client should be initialized");
+    };
```

Comment:
bit of nit but these can be combined:
```suggestion
    let (Some(bootstrapped_client), Some(client_config)) = (ctx.bootstrapped_client, ctx.arti_client_config) else {
        panic!("Arti client should be initialized");
    };
```

Reply:
well then the comments should be different and mention the field IMO